File: modules/07-scheduler/kotlin/lib/Sched.kt
// Kotlin/JVM parallelism here uses platform threads (1:1 OS threads). Kotlin
// also has coroutines (M:N, cooperative, the closest analog to goroutines) and,
// on modern JDKs, virtual threads (Project Loom) — both need dependencies or a
// newer runtime and are noted in the comparison; a CPU sweep uses plain threads.

import kotlin.math.sqrt

fun chunkSum(
    lo: Long,
    hi: Long,
): Double {
    var s = 0.0
    var i = lo
    while (i < hi) {
        s += sqrt(i.toDouble())
        i++
    }
    return s
}

fun parallelSqrtSum(
    total: Long,
    workers: Int,
): Double {
    val chunk = total / workers
    val results = DoubleArray(workers)
    val threads =
        (0 until workers).map { k ->
            Thread { results[k] = chunkSum(k * chunk, (k + 1) * chunk) }
        }
    threads.forEach { it.start() }
    threads.forEach { it.join() }
    return results.sum()
}

File: modules/07-scheduler/ocaml/lib/sched.ml
(* OCaml 5 parallelism is via DOMAINS: 1:1 with OS threads (unlike Go's M:N
   goroutines). Structured concurrency and cooperative I/O concurrency come from
   Eio (effect-based), a separate library; for a pure CPU sweep, domains are the
   direct comparison. *)

let chunk_sum lo hi =
  let s = ref 0.0 in
  for i = lo to hi - 1 do
    s := !s +. sqrt (float_of_int i)
  done;
  !s

let parallel_sqrt_sum total workers =
  let chunk = total / workers in
  let domains =
    List.init workers (fun k ->
        Domain.spawn (fun () -> chunk_sum (k * chunk) ((k + 1) * chunk)))
  in
  List.fold_left (fun acc d -> acc +. Domain.join d) 0.0 domains
